x-pack/legacy/plugins/siem/public/pages/detection_engine/rules/edit/index.tsx

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  EuiButton,
+  EuiCallOut,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiSpacer,
+  EuiTabbedContent,
+  EuiTabbedContentTab,
+} from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n/react';
+import React, { memo, useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { Redirect, useParams } from 'react-router-dom';
+
+import { HeaderPage } from '../../../../components/header_page';
+import { WrapperPage } from '../../../../components/wrapper_page';
+import { SpyRoute } from '../../../../utils/route/spy_routes';
+import { DETECTION_ENGINE_PAGE_NAME } from '../../../../components/link_to/redirect_to_detection_engine';
+import { useRule, usePersistRule } from '../../../../containers/detection_engine/rules';
+import { FormHook, FormData } from '../components/shared_imports';
+import { StepPanel } from '../components/step_panel';
+import { StepAboutRule } from '../components/step_about_rule';
+import { StepDefineRule } from '../components/step_define_rule';
+import { StepScheduleRule } from '../components/step_schedule_rule';
+import { formatRule } from '../create/helpers';
+import { getStepsData } from '../helpers';
+import * as ruleI18n from '../translations';
+import { RuleStep, DefineStepRule, AboutStepRule, ScheduleStepRule } from '../types';
+import * as i18n from './translations';
+
+interface StepRuleForm {
+  isValid: boolean;
+}
+interface AboutStepRuleForm extends StepRuleForm {
+  data: AboutStepRule | null;
+}
+interface DefineStepRuleForm extends StepRuleForm {
+  data: DefineStepRule | null;
+}
+interface ScheduleStepRuleForm extends StepRuleForm {
+  data: ScheduleStepRule | null;
+}
+
+export const EditRuleComponent = memo(() => {
+  const { ruleId } = useParams();
+  const [loading, rule] = useRule(ruleId);
+  const [initForm, setInitForm] = useState(false);
+  const [myAboutRuleForm, setMyAboutRuleForm] = useState<AboutStepRuleForm>({
+    data: null,
+    isValid: false,
+  });
+  const [myDefineRuleForm, setMyDefineRuleForm] = useState<DefineStepRuleForm>({
+    data: null,
+    isValid: false,
+  });
```

Comment:
It may be worthwhile to later refactor this to `useReducer` with the amount of `useState` that's being used here.

Reply:
10:4